### DISCUSSION TITLE
a couple of stupid questions

### DISCUSSION THREAD

**Discussion by j2969719** [rikyoz/bit7z#148] [Q&A] [answered]
is it possible to add only a directory item (w/o files) to a specific path in the archive?

how to correctly compress symlinks, especially those pointing to directories?

**Comment by rikyoz** [comment]
Hi!

> is it possible to add only a directory item (w/o files) to a specific path in the archive?

Bit7z allows adding only items existing on the filesystem, so you cannot add an empty non-existing directory. I might add this feature in the next versions of the library, though!

If you plan to add some files to the new directory inside the archive, you might achieve something similar by specifying a custom path, e.g.:

```cpp
    Bit7zLibrary lib{ "7z.dll" };
    BitArchiveWriter writer{ lib, BitFormat::SevenZip };
    writer.addFile( "path/to/file.txt", "new-directory/file.txt" );
    writer.compressTo( "output.7z" );
```

In this way, you're not directly creating an empty directory item, but the resulting archive will contain a directory with a custom name that will contain the files.

> how to correctly compress symlinks, especially those pointing to directories?

You should be able to add symbolic links via the `BitFileCompressor::compress` or the `BitArchiveWriter::addItems` methods.
However, I never specifically tested the compression of symlinks.
Did you have any issues while compressing symbolic links?

**Reply by j2969719** [reply]
in my case, both directories and files located in them passed to be compressed
initially used only addFile, but in this case dates/attributes and empty directories were lost
at the moment I use addItems, but since when adding a directory, files are also added - duplicates appear in the archive even with setUpdateMode(UpdateMode::Update);, so for now i simply exclude non-empty directories

addItems treats symlinks on directories as simple directories
in the archives created by bit7z, instead of the target path of the symlink to the file, for some reason libarchive get the contents of the file itself, although this may be on my side.

**Reply by rikyoz** [reply, marked as answer]
> initially used only addFile, but in this case dates/attributes and empty directories were lost

`addFile` ignoring empty directories is the expected behavior; on the other hand, this method should preserve the dates/attributes of the file being added, so I'll need to investigate this.

> at the moment I use addItems, but since when adding a directory, files are also added

Yeah, it's the expected behavior.

> duplicates appear in the archive even with setUpdateMode(UpdateMode::Update);

This is probably due to the fact that the duplicates check is performed only in the case of updating an existing archive, and it considers only duplicates between the new items being added and the old ones already present in the archive. The library doesn't check for duplicates among the new items.

> addItems treats symlinks on directories as simple directories
in the archives created by bit7z, instead of the target path of the symlink to the file, for some reason libarchive get the contents of the file itself, although this may be on my side.

Which archive format are you using when compressing symlinks?
I did some quick tests, and bit7z seems to behave the same as the 7z CLI program, at least when compressing using the 7z format. Actually, bit7z behaved better since it preserved the symbolic link attribute, while the 7z command didn't.

**Reply by j2969719** [reply]
tnx for the quick response

about addFile, meant that in this case the directories passed for compression had to be skipped and, as a result, their dates/attributes were not  preserved

sorry my bad, as a linux user, im just so used to the behavior of p7zip and libarchive that didnt even bother to check how 7zz behaves with symlinks

**Comment by j2969719** [comment]
also what is the best way to handle the situation when bitarchivereader received an archive with an encrypted header if the password was not set in advance?

**Reply by rikyoz** [reply]
I think the best way to handle this case is to catch the exception, request the password from the user, and retry opening the file with the given password

**Comment by j2969719** [comment]
hi again

is it possible to extract multiple items in predefined paths, but not into memory or a single output directory?  atm im using extraction one by one in ostream, which is very slow with large solid archives.

**Reply by rikyoz** [reply]
Hi!
I'm sorry for not getting back to you sooner!

> is it possible to extract multiple items in predefined paths, but not into memory or a single output directory? atm im using extraction one by one in ostream, which is very slow with large solid archives.

Do you mean extracting each item to possibly different output paths?

If you're using `BitFileExtractor`, part of the slow performance might be because the archive is opened each time you extract something from it.

Alternatively, you could use the `BitArchiveReader` class: it provides an `extractTo( const tstring& out_dir, const std::vector< uint32_t >& indices )` method, which you might call multiple times to extract the items in separate directories. Using this class, you open the archive only once.

Unfortunately, at the moment, bit7z doesn't provide a way to extract multiple items in predefined paths in one step, but I'll definitely evaluate how to provide such an API!

**Reply by j2969719** [reply]
yep, tried this method, but in my case it will require creating a temporary folder, extracting and moving to the desired location, but im too lazy so decided to ask if there is an easier solution maybe

thanks anyway and sorry for the hassle
